### PR TITLE
Use multiprocessing for the scheduler

### PR DESCRIPTION
**Pull request by bolkedebruin**
This parallelizes the scheduler to the amount of cpu-cores.

### CODE REVIEW

**Comment by criccomini** [comment]
Relates to https://github.com/airbnb/airflow/pull/1374

File: airflow/jobs.py [hunk 1]
```diff
@@ -422,7 +424,7 @@ def schedule_dag(self, dag):
 
             # don't ever schedule prior to the dag's start_date
             if dag.start_date:
-                next_run_date = max(next_run_date, dag.start_date)
+                next_run_date = max(next_run_date or dag.start_date, dag.start_date)
```
Comment (criccomini) [inline airflow/jobs.py]: Nice catch. Perhaps a little more intelligible (though more verbose) would be:

```
dag.start_date if not next_run_date else max(next_run_date, dag.start_date)
```

File: airflow/jobs.py [hunk 2]
```diff
@@ -657,6 +659,23 @@ def prioritize_queued(self, session, executor, dagbag):
 
                 session.commit()
 
+    def _split_dags(self, dags, size):
```
Comment (criccomini) [inline airflow/jobs.py:674]: Pydocs would be helpful. This stuff is getting a little gnarly.

File: airflow/jobs.py [hunk 3]
```diff
@@ -657,6 +659,23 @@ def prioritize_queued(self, session, executor, dagbag):
...
+        size = max(1, size)
+        return [dags[i:i + size] for i in range(0, len(dags), size)]
+
+    def _do_dags(self, dagbag, dags):
```
Comment (criccomini) [inline airflow/jobs.py]: Pydocs would be helpful. This stuff is getting a little gnarly.

File: airflow/jobs.py [hunk 4]
```diff
@@ -695,18 +714,19 @@ def _execute(self):
...
-                    dag = dagbag.get_dag(dag.dag_id)
-                    if not dag or (dag.dag_id in paused_dag_ids):
-                        continue
-                    try:
-                        self.schedule_dag(dag)
-                        self.process_dag(dag, executor)
-                        self.manage_slas(dag)
-                    except Exception as e:
-                        self.logger.exception(e)
+                dags = [x for x in dags if x.dag_id not in paused_dag_ids]
```
Comment (criccomini) [inline airflow/jobs.py:738]: I think you can use [filter()](https://docs.python.org/2/library/functions.html#filter) to this a little more pythonically.

File: airflow/jobs.py [hunk 5]
```diff
@@ -695,18 +714,19 @@ def _execute(self):
...
+                self.logger.info("CPU: {} DAGs:{}".format(multiprocessing.cpu_count(), len(dags)))
```
Comment (criccomini) [inline airflow/jobs.py]: Nit: "Total CPUs: {}" would be a little more intelligible. At first, I thought this was saying "I am CPU number N, I'm going to process DAGs 1-10"

Comment (criccomini) [inline airflow/jobs.py]: How expensive is cpu_count? I notice we're calling it multiple times in the scheduler loop.

File: airflow/utils/db.py [hunk 6]
```diff
@@ -83,6 +83,23 @@ def merge_conn(conn, session=None):
         session.commit()
 
 
+@event.listens_for(settings.engine, "connect")
```
Comment (criccomini) [inline airflow/utils/db.py:86]: What are these? MySQLAlchemy foo is lacking.

**Comment by criccomini** [comment]
Overall, looks good. A few comments:
1. I think @mistercrunch should have a look at this.
2. I think we might want a num_cores config in the scheduler to define the max number of cores to use. Right now it looks like the PR uses all cores on the box, always.

File: airflow/jobs.py (hunk 4 above)
Comment (bolkedebruin) [inline airflow/jobs.py:738]: with a lambda? I personally don't like to read that stuff so much

File: airflow/utils/db.py (hunk 6 above)
Comment (bolkedebruin) [inline airflow/utils/db.py:86]: It basically checks when a new connecting is made. You cannot use the same db connection in a fork so we force the db logic to make a new connection when it detects that the pid of the process that created the connection is not the same as it is currently in.

File: airflow/jobs.py (hunk 5 above)
Comment (bolkedebruin) [inline airflow/jobs.py]: don't know to be honest. I haven't seen any impact though

File: airflow/configuration.py [hunk 7]
```diff
@@ -414,6 +420,7 @@ def run_command(command):
 job_heartbeat_sec = 1
 scheduler_heartbeat_sec = 5
 authenticate = true
+max_threads = 2
```
Comment (r39132) [inline airflow/configuration.py:423]: Nit  : maybe we should call this max_proc?

File: airflow/jobs.py [hunk 8]
```diff
@@ -422,7 +435,7 @@ def schedule_dag(self, dag):
 
             # don't ever schedule prior to the dag's start_date
             if dag.start_date:
-                next_run_date = max(next_run_date, dag.start_date)
+                next_run_date = dag.start_date if not next_run_date else max(next_run_date, dag.start_date)
```
Comment (r39132) [inline airflow/jobs.py:438]: This is an important bug fix :-) Should be a stand-alone commit or PR.

It looks like the cause of this is that a `TaskInstance` has not yet been inserted for the first run. As a result, even though `last_scheduled_run` is `None`, then `latest_run` is also `None`.  If `latest_run` is `None`, then the logic above picks the min start date for all of the tasks in the dag. This should never be None -- this is where there might be a bug. 

The logic flow above, for others looking at this code for the first time is 
1) look at dagruns (which the scheduler inserts into - `airflow run` and `airflow backfill` do not currently insert into this table) 
2) look at task_instance (which is presumably inserted after dag_runs) 
3) look at the dag bag for a start date

It seems counterintuitive that the tasks in the dagbag would not have a start date

File: airflow/configuration.py (hunk 7 above)
Comment (bolkedebruin) [inline airflow/configuration.py:423]: I was thinking "threads" is better understood as a concept. The executor uses `parallelism` but that is also a bit unclear.

File: airflow/jobs.py [hunk 9]
```diff
@@ -668,8 +706,8 @@ def _execute(self):
         dagbag = models.DagBag(self.subdir, sync_to_db=True)
         executor = self.executor = dagbag.executor
         executor.start()
-        i = 0
-        while not self.num_runs or self.num_runs > i:
+        self.runs = 0
```
Comment (jlowin) [inline airflow/jobs.py:709]: Is there an advantage to making this `self.runs` instead of just `runs`?

Comment (bolkedebruin) [inline airflow/jobs.py:709]: Yes, that it actually counts :-).  When I kept it a local variable it would not properly increase. I am not sure why.

File: airflow/jobs.py (hunk 8 above)
Comment (bolkedebruin) [inline airflow/jobs.py:438]: Ok it is a separate commit now. Real fix should be in another PR I guess.

**Comment by mistercrunch** [comment]
Oh sweet, I just realized this was happening! Amazing. This should help speed up or schedule cycles that have been slipping lately from <1minute to 7+ minutes

**Comment by bolkedebruin** [comment]
@mistercrunch Thanks! You might have been hit by #1374 as well, the fix also improved speed quite a lot for @criccomini

**Comment by bolkedebruin** [comment]
@r39132 Yes and no. I want to have a look at that separately and understand it well, therefore first the partial fix and then I want to address the more core issue in a new PR.
